pennylane/collections/metric_tensor.py

```diff
@@ -0,0 +1,292 @@
+            of devices of length corresponding to the number of parametric layers in ``qnode``.
+    """
+
+    # pylint: disable=too-many-instance-attributes
+
+    def __init__(self, qnode, device):
+        self.qnode = qnode
+        self.dev = device
+        self.interface = qnode.interface
+        self._qnode_type = self.qnode.__class__
+
+        self.qnodes = None
+        self.obs = None
+        self.coeffs = None
+        self.params = None
+
+        if self.interface == "tf":
+            import tensorflow as tf
+
+            self.np = tf
+        elif self.interface == "torch":
+            import torch
+
+            self.np = torch
+        elif self.interface == "autograd":
+            from autograd import numpy as np  # pylint: disable=redefined-outer-name
+
+            self.np = np
+        else:
+            self.np = np
+
+        self.dot = _get_dot_func(self.interface)[0]
+
+    def __call__(self, *args, **kwargs):
+        parallel = kwargs.pop("parallel", False)
+        diag_approx = kwargs.pop("diag_approx", False)
+
+        if self.qnodes is None:
+            self.qnodes, self.obs, self.coeffs, self.params = self._make_qnodes(args, kwargs)
+
+        probs = self.qnodes(*args, parallel=parallel)
+        self.np.reshape(probs, [len(self.qnodes), -1])
+
+        gs = []
+
+        for prob, obs, coeffs in zip(probs, self.obs, self.coeffs):
+            # calculate the covariance matrix of this layer
+            scale = np.outer(coeffs, coeffs)
+
+            if self.interface == "torch":
+                scale = self.np.tensor(scale)  # pylint: disable=not-callable
+
+            g = scale * self.cov_matrix(prob, obs, diag_approx=diag_approx)
+            gs.append(g)
+
+        perm = np.array([item for sublist in self.params for item in sublist])
+
+        if self.interface == "torch":
+            # torch allows tensor assignment
+            tensor = self.np.zeros([len(perm), len(perm)], dtype=self.np.float64)
+
+            for g, p in zip(gs, self.params):
+                row = np.array(p).reshape(-1, 1)
+                col = np.array(p).reshape(1, -1)
+                tensor[row, col] = g
+
+            return tensor
+
+        if self.interface == "tf":
+            # tf doesn't allow tensor assignment, but does provide
+            # support for constructing block diagonal operators
+
+            tfl = self.np.linalg
+            linop_blocks = [tfl.LinearOperatorFullMatrix(block) for block in gs]
+            linop_block_diagonal = tfl.LinearOperatorBlockDiag(linop_blocks)
+
+            # permute rows
+            tensor = self.np.gather(linop_block_diagonal.to_dense(), perm)
+
+            # permute columns
+            tensor = self.np.gather(self.np.transpose(tensor), perm)
+            return tensor
+
+        # autograd
+        tensor = block_diag(*gs)[:, perm][perm]
+        return tensor
+
+    def cov_matrix(self, prob, obs, diag_approx=False):
```

Comment:
Could we add your covariance matrix code that has it as a QNodeCollection to the codebase here? In this way we could also support non-commuting generators, couldn't we? One could think of keeping this code as `cov_matrix_commuting` (because it is just cheaper to evaluate)

Reply:
Yep! This could be refactored to move the `cov_matrix` method as a class in a separate `cov_matrix.py` file.

Reply:
This should definitely be on the roadmap!